### ISSUE TITLE
Multiple detection records for tags detected on Lotek receivers (pre and post UTC correction)

### ISSUE THREAD

**Issue by jsayers-bsc** [MotusDev/Motus-TO-DO#868]
Project 279 recently inquired about apparent duplicate records in their data (as seen in the alltags view). The records are from same batches and, to them anyway, also have identical timestamps. However they are using the ts field and the detections are from Lotek receivers which have apparently had a UTC offset applied. So there are now two records for each of these -- one where tsCorrected is the same as ts and the other which has had an offset applied. I assume that the non-corrected record shouldn't also be here...

**Comment by jsayers-bsc** [comment]
![image](https://user-images.githubusercontent.com/55001144/99708037-460c2380-2a6b-11eb-866a-3f5876b69968.png)

**Comment by jsayers-bsc** [comment]
Relatedly, how the R Package handle records that have been modified/ are newer in the source data than in the local database?

Do changes to tsCorrected fall under the category "metadata" that gets updated when foreMeta = T?

**Comment by steffilazerte** [comment]
Two points:

1. If the `hitID`s are unique (as they are here), these duplications are arriving from the server (the R package doesn't create IDs) - So a problem for @denislepage and @cjardine-bsc  :grin:

2. In answer to @jsayers-bsc's question:
    - `utcOffset` and `ts` are the fields used in `alltags` and `alltagsGPS` views to calculate `tsCorrected`
    - `utcOffset` is part of `recvDeps` which gets updated as metadata
    - `ts` is part of `hits` which would not be updated because those updates are based on the last batchID downloaded
    - So in conclusion, `tsCorrected` isn't metadata, but would be updated if the metadata used to calculate it gets updated

I'm unassigning myself and removing the R package label from this because I think this is a server issue, not a package issue (but happy to change that decision if warranted!)

**Comment by denislepage** [comment]
Have we ever set up the process for deprecating batches? I know we had a plan for it. While we remove hits for batches that have been re-run from our server, those will remain on the user end, and new ones will be added. As a minimum, we should supply a list of all the deprecated batches, and update the batch table with a deprecated flag. There should also be a function on the R side that allows compacting the database to remove these obsolete batches, but this should be done by users when needed.

**Comment by steffilazerte** [comment]
We have talked about it for a while but haven't implemented anything. This is the time of year when I have more free time, so now would be a great time to do this! But I think best to open a new issue/rename this issue so we can keep track of it and make sure we don't forget again.

**Comment by cjardine-bsc** [comment]
Cool. We have three R views set up that provide a list of deprecated batches.

They are basically lists of batchIDs that can be flagged as deprecated and/or removed. 

vwR_Batches_for_all_deprecated
vwR_Batches_for_receiver_deprecated
vwR_Batches_for_tag_project_deprecated

@pmorrill Do already have API entry points for these? If not will you create them?

**Comment by cjardine-bsc** [comment]
@jsayers-bsc Back to your original questions. Those aren't duplicates (look at sig, sigsd, etc). They are unique hits. I suspect those ts values are actually different, but what is displayed there is just the date, hour and min. The values would differ from one another only by a few seconds. 

So the TS should be different when viewed as an integer.

**Comment by steffilazerte** [comment]
Or did I jump the gun and should we open a new issue?

**Comment by cjardine-bsc** [comment]
We should find the issue that I'm sure already exists for that. lol

**Comment by jsayers-bsc** [comment]
> @jsayers-bsc Back to your original questions. Those aren't duplicates (look at sig, sigsd, etc). They are unique hits. I suspect those ts values are actually different, but what is displayed there is just the date, hour and min. The values would differ from one another only by a few seconds.
> 
> So the TS should be different when viewed as an integer.

Yes I was just reading back from the beginning to refresh my memory and noticed this!

**Comment by pmorrill** [comment]
> 
> 
> Cool. We have three R views set up that provide a list of deprecated batches.
> 
> They are basically lists of batchIDs that can be flagged as deprecated and/or removed.
> 
> vwR_Batches_for_all_deprecated
> vwR_Batches_for_receiver_deprecated
> vwR_Batches_for_tag_project_deprecated
> 
> @pmorrill Do already have API entry points for these? If not will you create them?

They exist:

/batches_for_all_deprecated
/batches_for_receiver_deprecated
/batches_for_tag_project_deprecated

**Comment by steffilazerte** [comment]
So just up to me to add this functionality to R package sandbox then? I'll take a look. @denislepage @cjardine-bsc, did you have previously thought out plans for how this should be implemented, or should I see what I think would be best?

**Comment by denislepage** [comment]
The first thing I would do is make sure the motusFilter field is available in the batches table schema, so existing batches can be flagged. Paul and Catherine shared some API entry points earlier.

Anything in the returned list should be used to update the batches table accordingly.

Deleting records from the database is a little more complex, especially because of the disconnect between batches and runs. There would ideally be a separate script that loops through the batch table records that have been marked as deprecated during the previous step, and then deletes the database records associated with these batchID’s. I would likely do this in a loop. The more useful table to clean up is probably the bigger hits one.

We have a different status code in motusFilter for batches that are marked as deprecated (=2), and those that have been cleaned up (=4), and I suggest you use the same values for the R client.

The script could first get a list of all batchID’s from the batches table where batchFilter = 2.

Then loop through and deleted associated records from the various tables for individual batchID values. Not all tables should be deleted. I would at least make sure to keep the deprecated batches record. Deleting runs records would also need to do a join (runs may spread across several batches), but I’d need to look more carefully at the code. We could mainly worry about hitsDeployID and hits for now for testing. Runs would probably also be important to remove.

Finally, we should probably send a VACUUM call to compact the database and reclaim the deleted space. Perhaps this should be an optional feature. We’ll need to test how slow this might be on a very large db.

**Comment by cjardine-bsc** [comment]
Just to clarify @steffilazerte Denis is talking about the batchFilter column, not the motusFilter column (that a different beastie). 

I've added batchFilter to the three R views. It is almost always 4 (deleted) not 2 (deprecated (re-run) because batches are deleted periodically and updated from 2 to 4. So 2 is more of a holding tank for newly re-run batches. We also have a batch filter of 6 for cases where the boot number changed, those have been manually deprecated. And a filter of 5 for batches that are currently being deleted. 

I doubt we need to make all those distinctions clear to users. Perhaps instead of the actual batchFilter value we can just use a Y/N T/F boolean in R. Anything returned by the three entry points noted above should be removed. 

Just to confirm for @denislepage et al. det_batches_projects records are not being deleted as part of the batch scrubbing process. So we're good to go.

**Comment by denislepage** [comment]
I think the batchFilter (yup) field needs at least 3 values: valid, marked for deprecation and deleted. We already have numeric values for these concepts, though users should have a simple way to know what they mean (e.g. that works well if any batchFilter > 0 means a batch to be ignored).

The process that marks the batches as deprecated should be the one that calls the API.

Deleting the data for deprecated batches should be a separate function not relying on the API itself, but only on the filter value from the table. Once a batch is deleted, then the filter value would be updated from deprecated to deleted.

Also, it’s worth noting that at the moment, we are not tracking the datetime when a batch is deprecated, so every API call will receive the full list of deprecated batchID’s for the project or receiver (e.g., we can’t currently ask for a list of newly deprecated batches).

Some of those batches received may already be marked in the local database as deprecated or deleted (and other batchID’s may not exist locally). We should make sure that only batches currently marked as valid are changed to deprecated. Otherwise, deleted batches may be reverted back to deprecated, which will force the delete process to try and delete them all needlessly.

**Comment by cjardine-bsc** [comment]
Ahhh, That means the batchFilter in the R package will be different than the one in our DB, thats fine. We maybe don't need to return batchFilter in the API then. It will be set in R.

**Comment by steffilazerte** [comment]
@pmorrill I've got `batches_for_tag_project_deprecated` and `batches_for_receiver_deprecated` working, but when I try `batches_for_all_deprecated` with a `batchID` of 0 (JSON boxed or unboxed) I always receive the error "You do not have permission to access the data for this project or receiver".

I assumed that this would be an 'open' entry point (like activity_for_all or gps_for_all), but perhaps it is it an admin entry point like batches_for_all, where I wouldn't have access?

Any ideas of what's going on? Am I doing something wrong? Is this entry point not yet set up?

**Comment by steffilazerte** [comment]
Okay, I've started working this out @denislepage and I have a couple of questions:

**1. Optional?**
I would argue that this shouldn't be an optional thing, that anytime the user updates their motus SQLite file, deprecated batches should be marked and removed. Is this reasonable? Or will that be really slow on large databases, so it's better to have a separate remove deprecated batches function which users can run manually?
  - If we can run this with tagme in one step (so deprecated batches are immediately removed), I'll only mark batches as 'deprecated'
  - Otherwise I'll mark them as 'deprecated' or 'deprecated/deleted' depending on their status

**2. Removing data**
I would mark batchIDs in the `batches` table, but would remove them everywhere else (hits, activity, gps, runs, etc.)

**3. Removing deprecated batches from the runs table**
If I understand correctly, a run can start and end in one batch, or can start in one batch and end in another. So I can't simply remove runs that *start* in a deprecated batch.
I assume a good way to deal with this would be to remove hits in deprecated batches and then to remove runs that are no longer part of the hits table (i.e. if all the hits in a particular run were removed). Does that make sense?
What happens to the runs start batchID if the batch is deprecated? Does it get a new start batchID?

**Comment by pmorrill** [comment]
The 'batches_for_all function is currently restricted to admin users, probably because it dumps data for all project receivers....which normally would not be available to non-admin users. @denislepage can confirm that.

**Comment by pmorrill** [comment]
The batches_for_all_deprecated func is similar. I could adjust it so that it allows 'sgAdmin' users to get data (I presume @steffilazerte falls into that class), but it would not help with other users.

**Comment by steffilazerte** [comment]
I don't think other users need access, but I can't test that the functions work in R if `testuser` doesn't have access. Are these entry points used by others? If no admins use them in R, an alternative would be to just not worry about creating functions to access them in R.

**Comment by denislepage** [comment]
I am almost positive the batches_for_all function isn't needed from R. This is a legacy part of the code. We do use batches_for_all when pulling data from the sgdata server, so we can iterate through all batches across projects. Initially the API for the R client on motus and the one on sgdata were created to be the same, but they now serve different functions. I would likely even be inclined to disable the batches_for_all entry point in Motus (but not in sgdata, obviously).

**Comment by steffilazerte** [comment]
Okay, so I won't use the batches_for_all_deprecated and will probably remove batches_for_all as well.

@denislepage, can you review my [previous comment](https://github.com/MotusDev/Motus-TO-DO/issues/868#issuecomment-870853628) and let me know what you think?

**Comment by steffilazerte** [comment]
I've got some preliminary work on this but can't see to find any projects that actually have deprecated batches.

`batches_for_tag_project_deprecated` returns `batchID`s and `batchFilter`s but all the filter values are 4, which I think means cleaned up. I haven't found a single 2 (deprecated). 

Am I missing something? Or are there really no deprecated batches recorded? (I looped through all the project numbers).

**Comment by StuMackenzie** [comment]
yeah, sorry for the misunderstanding, just syntax in a way. I was wondering this and what the real potential implications could be...

presumably we're making tweaks to reprocessed data from time to time so it would be good to think that data would always be better than what was present before, but I suppose we can't guarantee that.

**Comment by cjardine-bsc** [comment]
> 
> 
> I just had a quick look. Of the 132 batches in the sample data, 7 have been deprecated. I have created a new table that contains those 7 records in sandbox, @steffilazerte if you want to try that on the sample project.
> 
> The only other thing that might be useful is to add the 7 batches that were created in replacement to the sample table, as would normally happen in real life. That would allow to provide some examples that illustrates how the database may end up with duplicate records.
> 
> If we wanted to add those to the sample data, I think this gives the list of new batches that cover the same as the deprecated ones on the same receiver. but it returns 8 instead of 7, so it may not be a perfect fit (though it's no guaranteed that replacement batches will cover exactly the same periods).
> 
> The total number of records in the deprecated batches vs. the new ones below is also pretty close (92949 vs. 92516), so it may be close enough. Again, they are not expected to match perfectly.
> 
> ```
> select a.* from det_batches a
> inner join Sample_R_Batches_for_tag_project_deprecated b on a.motusDeviceID = b.motusDeviceID 
> inner join det_batches c on b.batchID = c.batchID
> and a.batchID <> c.batchID
> and (a.tsBegin <= c.tsEnd) and (a.tsEnd >= c.tsBegin)
> and a.batchFilter = 0
> ```

Hey D-Boss, 

I'm confused. I see 135 batches in the sample data. And all of them have been deprecated. 

SELECT *
  FROM [sample_r_batches_for_tag_project] a inner join det_batches b on a.batchID = b.sgbatchID 
  where batchFilter > 0 
  
Remember the batchIDs in the sample data match the sgBatchIDs in the det_batches table. Though even changing that join doesn't explain where this list of batches came from. Am I missing something?

Regardless. It doesn't really matter, we could just choose an arbitrary number of batches to list as deprecated. Are you happy with the seven you picked? If so I'll add the data from their replacement batches to the sample tables.

**Comment by denislepage** [comment]
Ah, yes, I likely failed to join by sgBatchID.

Anyway, picking a few of them arbitrarily makes sense. Not sure if my 7 are better or worse than others. Picking a few that come from the same receiver is probably more logical, but perhaps sample data is all from the same anyway.

I guess the API doesn’t filter out the deprecated batches then in the sample table. That filtering is perhaps done in the view, not the API call to the db.

**Comment by cjardine-bsc** [comment]
KK. I'm choosing **closes eyes and points at table** device ID 821! 

I've added all the batches, runs and hits data for the replacement batches (n=6)  to the sample tables in sandbox and live. 

Note that the batchIDs are not the same between sandbox and live.

**Comment by steffilazerte** [comment]
Awesome! This makes the examples and tests so much easier! It'll also be much easier to explain/explore in the documentation.

I'm going to leave this issue open until I get a chance to add the documentation, just in case 😉

**Comment by leberrigan** [comment]
@steffilazerte just checking if you want to keep this open.

**Comment by steffilazerte** [comment]
I still need to add documentation for this

**Comment by steffilazerte** [comment]
Functions to identify and remove deprecated batches are in the current version of the motus package (above)

I have also created an article explaining how/why to remove deprecated batches (https://motuswts.github.io/motus/articles/deprecated-batches.html). 

@ajmacdonald3, @StuMackenzie, @j-sayers, can one of you take a quick look and let me know if you see any problems?

Once someone reviews that, I'll officially add the article to the documentation and close this issue.

**Comment by steffilazerte** [comment]
Unless I hear back to the contrary, I'll assume that in one week, I can just add this to the docs and close the issue :grin:

**Comment by steffilazerte** [comment]
Thanks @ajmacdonald3, no problem about the delay. I'm happy to wait until you get a chance to try it out, I just didn't want to put it off forever :grin: 

Otherwise I'll merge it in next week, thanks!

**Comment by ajmacdonald3** [comment]
Sorry for taking so long, @steffilazerte, I have finally tested the functions for removing deprecated batches in R. They worked well for me on both project 176 and project 349.

**Comment by steffilazerte** [comment]
Thanks, no problem! I'll merge in the docs
